node-graph/gcore/src/vector/vector_data/attributes.rs

```diff
@@ -95,9 +95,25 @@ impl PointDomain {
 		self.positions.clear();
 	}
 
-	pub fn retain(&mut self, f: impl Fn(&PointId) -> bool) {
+	pub fn retain(&mut self, segment_domain: &mut SegmentDomain, f: impl Fn(&PointId) -> bool) {
 		let mut keep = self.id.iter().map(&f);
 		self.positions.retain(|_| keep.next().unwrap_or_default());
+
+		let mut id_map = Vec::with_capacity(self.ids().len());
+		let mut new_index = 0;
+		for id in self.ids() {
+			if f(id) {
+				id_map.push(new_index);
+				new_index += 1;
+			} else {
+				id_map.push(usize::MAX); // A placeholder for invalid ids. This is checked after the segment domain is modified.
+			}
+		}
+
+		let update_index = |index: &mut usize| *index = id_map[*index];
+		segment_domain.start_point.iter_mut().for_each(update_index);
+		segment_domain.end_point.iter_mut().for_each(update_index);
+
```

Comment:
By building a prefix sum over the keep vector we should be able to avoid this allocation. But we can do that when benchmarking shows that this is a bottleneck